p256/fiat_p256.mli

```diff
@@ -1,61 +1,31 @@
-(** A point on the P-256 curve (public key), that is not the point at infinity. *)
-type point
-
-(** The type for point parsing errors. *)
-type point_error =
+(** The type for public key parsing errors. *)
+type error =
   [ `Invalid_range
   | `Invalid_format
   | `Invalid_length
   | `Not_on_curve
   | `At_infinity ]
 
-val pp_point_error : Format.formatter -> point_error -> unit
-(** Pretty printer for point parsing errors *)
-
-val point_of_cs : Cstruct.t -> (point, point_error) result
-(** Convert from cstruct. Refuses the point at infinity.
-    The format is the uncompressed format described in
-    SEC1, section 2.3.4, that is to say:
-
-    - the point at infinity is the single byte "00".
-    - for a point (x, y) not at infinity, the format is:
-      - the byte "04"
-      - x serialized in big endian format, padded to 32 bytes.
-      - y serialized in big endian format, padded to 32 bytes.
-
-    @see <http://www.secg.org/sec1-v2.pdf>
-*)
-
-val point_of_hex : Hex.t -> (point, point_error) result
-(** Convert a point from hex. See [point_of_cs]. *)
-
-val point_to_cs : point -> Cstruct.t
-(** Convert a point to a cstruct. See [point_of_cs] for the format. *)
-
-(** A scalar value. *)
-type scalar
+val pp_error : Format.formatter -> error -> unit
+(** Pretty printer for public key parsing errors *)
 
-(** The type for scalar parsing errors. *)
-type scalar_error =
-  [ `Invalid_length
-  | `Invalid_range ]
+(** Type for P256 private keys *)
+type secret
 
-val pp_scalar_error : Format.formatter -> scalar_error -> unit
-(** Pretty printer for scalar parsing errors *)
+val gen_key : rng:(int -> Cstruct.t) -> secret * Cstruct.t
+(** [gen_key ~rng] generates a private and a public key for Ephemeral Diffie-Hellman
+    over P256. The returned key pair MUST only be used for a single key exchange.
 
-val scalar_of_cs : Cstruct.t -> (scalar, scalar_error) result
-(** Read data from a cstruct.
-    It should be 32 bytes long, in big endian format. Returns an error when the
-    number is zero, or if it is larger than or equal to the group order. *)
+    [rng] is the function used to repeteadly generate a private key until a valid candidate
+    is obtained. [rng]'s int parameter is the size of the [Cstruct.t] to generate.
+    If [rng] returns an invalid length buffer, [Failure _] is raised.
 
-val scalar_of_hex : Hex.t -> (scalar, scalar_error) result
-(** Like [scalar_of_cs] but read from hex data. *)
+    The generated private key is checked to be greater than zero and lower than the group
+    order meaning the public key cannot be the point at inifinity. *)
```

Comment:
please include a comment about the second projection of the returned value -- is it truncated if there are leading 0 bytes? (i suspect no, which is fine (needed in TLS 1.3, whereas truncation is require in TLS 1.2 and earlier)).

Reply:
I'm not sure which truncation you're talking about? It shouldn't matter for the `secret` as it's an abstract type so I'm guessing you mean for the returned encoded public key in the Cstruct but I'm unsure what there is to truncate there.

Reply:
sorry, its too warm -- i was thinking about https://tools.ietf.org/html/rfc5246#section-4.4 -- but it looks like mainly leading 0 bytes of the computed shared secret Z are stripped (see https://tools.ietf.org/html/rfc5246#section-8.1.2) (that is tls 1.2, 1.3 is different in that behaviour).

Reply:
No worries ! I updated the `key_exchange` documentation according to your other similar comment !